### PR TITLE
CosmosDB Restoreable APIs for SQL and MongoDB database account

### PR DESCRIPTION
**Pull request by vinhms**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/vinhms/azure-rest-api-specs/dev-cosmos-db-Microsoft.DocumentDB-2020-06-01-preview...master/).

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Review by kavskalyan — COMMENTED** [review]
LGTM

**Comment by chiragg4u** [comment]
Please go through the contribution checklist added as part of the review and mark them completed. Also Please go through the ARM API Review Checklist. These steps are required before review can be continued.

**Comment by vinhms** [comment]
I have gone through the checklists and marked them appropriately.  @chiragg4u can you let me know what the next step is?  looks like it's not being auto-assigned to anyone.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 1]
```diff
@@ -7820,25 +8644,69 @@
...
+              "type": "string",
+              "readOnly": true,
+              "description": "The name of this restorable MongoDB collection."
+            },
+            "ownerResourceId": {
+              "type": "string",
+              "readOnly": true,
+              "description": "The resource Id of this restorable MongoDB collection."
+            },
+            "documentCollection": {
```
Comment (vinhms) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:8678]: "collection" instead of "documentCollection"

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 2]
```diff
@@ -7695,123 +8139,503 @@
...
-          "enum": [
-            "Periodic",
-            "Continuous"
-          ],
-          "description": "Describes the mode of backups.",
-          "x-ms-enum": {
-            "modelAsString": true,
-            "name": "BackupType"
+        "resource": {
+          "properties": {
```
Comment (kavskalyan) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:8383]: Add "type" for resource?

**Comment by vinhms** [comment]
DatabaseRestoreResource should be api specific

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 3]
```diff
@@ -4298,6 +4298,450 @@
...
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/locationParameter"
+          },
+          {
+            "$ref": "#/parameters/instanceIdParameter"
+          },
+          {
+            "$ref": "#/parameters/restorableSqlDatabaseRidParameter"
```
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4409]: A collection GET with a required query parameter is not going to be consumable by downstream services like ARG and Policy.  Why was this moved from a child resource of restorableSqlDatabases to its own GET with a required query param?  At minimum the query param should be optional and show containers from all databases

Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4409]: Thanks.. updated the spec to make these parameters optional

**Review by pilor — APPROVED** [review]

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 4]
```diff
@@ -4298,6 +4298,450 @@
           }
         }
       }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/locations/{location}/restorableDatabaseAccounts/{instanceId}/restorableSqlDatabases": {
+      "get": {
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4303]: Can we move API(sql/mongo) specific APIs to a new swagger file ? This way when we move from preview to stable the transition would be easier, plus the current file has grown immensely. You can refer the definitions from cosmos-db.json into the new swagger file.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{
+  "parameters": {
+    "api-version": "2020-06-01-preview",
+    "subscriptionId": "subid",
+    "location": "WestUS",
+    "instanceId": "98a570f2-63db-4117-91f0-366327b7b353",
+    "restorableMongodbDatabaseRid": "PD5DALigDgw="
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json:7]: I wonder if these are actual ids of some of the Prod resources. Please use all zeros in guids and in strings you can use: "myPropertyNameObject" (Same comment everywhere else (request/response))

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlResourceList.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    "restoreLocation": "WestUS",
+    "restoreTimestampInUtc": "10/13/2020 4:56"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "databaseName": "Database1",
+            "collectionNames": [
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlResourceList.json:16]: >collectionNames [](start = 13, length = 15)

nit: containerNames

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 7]
```diff
@@ -7688,130 +8132,492 @@
...
+      "description": "Enum to indicate the API type of the restorable database account.",
+      "type": "string",
+      "readOnly": true,
+      "enum": [
+        "MongoDB",
+        "Gremlin",
+        "Cassandra",
+        "Table",
+        "Sql",
+        "Etcd",
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:8311]: In my knowledge, we dont support it anymore. Let us refrain from exposing it here.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 8]
```diff
@@ -7688,130 +8132,492 @@
...
-            "modelAsString": true,
-            "name": "BackupType"
+        "resource": {
+          "properties": {
+            "_rid": {
+              "type": "string",
+              "readOnly": true,
+              "description": "A system generated property. A unique identifier."
+            },
+            "operationType": {
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:8388]: OperationType will have values from a fixed set ? If so, can we make it a Enum ?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseList.json [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+          {
+            "id": "/subscriptions/2296c272-5d55-40d9-bc05-4d56dc2d7588/providers/Microsoft.DocumentDb/locations/westus/restorableDatabaseAccounts/d9b26648-2f53-4541-b3d8-3044f4f9810d/restorableSqlDatabases/8456cb17-cdb0-4c6a-8db8-d0ff3f886257",
+            "type": "Microsoft.DocumentDB/locations/restorableDatabaseAccounts/restorableSqlDatabases",
+            "name": "8456cb17-cdb0-4c6a-8db8-d0ff3f886257",
+            "properties": {
+              "resource": {
+                "_rid": "ESXNLAAAAA==",
+                "eventTimestamp": "2020-09-02T19:53:42Z",
+                "ownerId": "Database1",
+                "ownerResourceId": "3fu-hg==",
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseList.json:43]: are these really required ? In every case, these values will be equal to database.id and database._rid. Why would we return two objects which represent the same concept.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseGet.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    "api-version": "2020-06-01-preview",
+    "subscriptionId": "2296c272-5d55-40d9-bc05-4d56dc2d7588",
+    "location": "WestUS",
+    "instanceId": "d9b26648-2f53-4541-b3d8-3044f4f9810d",
+    "restorableSqlDatabaseId": "59c21367-b98b-4a8e-abb7-b6f46600decc"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "id": "/subscriptions/2296c272-5d55-40d9-bc05-4d56dc2d7588/providers/Microsoft.DocumentDb/locations/westus/restorableDatabaseAccounts/36f09704-6be3-4f33-aa05-17b73e504c75/restorableSqlDatabases/59c21367-b98b-4a8e-abb7-b6f46600decc",
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseGet.json:12]: >36f09704-6be3-4f33-aa05-17b73e504c75 [](start = 142, length = 36)

this should be same as instanceId value in parameters

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        "value": [
+          {
+            "id": "/subscriptions/subid/providers/Microsoft.DocumentDb/locations/westus/restorableDatabaseAccounts/98a570f2-63db-4117-91f0-366327b7b353/restorableSqlContainers/79609a98-3394-41f8-911f-cfab0c075c86",
+            "type": "Microsoft.DocumentDB/locations/restorableDatabaseAccounts/restorableSqlContainers",
+            "name": "79609a98-3394-41f8-911f-cfab0c075c86",
+            "properties": {
+              "resource": {
+                "_rid": "zAyAPQAAAA==",
+                "eventTimestamp": "2020-10-13T04:56:42Z",
+                "ownerId": "Container1",
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json:21]: >ownerId [](start = 17, length = 7)

Can we rename owner to something else, it is not clearly indicative of its nature. How does something like, originalContainerId, sound.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbResourceList.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    "restoreTimestampInUtc": "10/13/2020 4:56"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "databaseName": "Database1",
+            "collectionNames": [
+              "Container1"
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbResourceList.json:17]: >              "Container1 [](start = 0, length = 25)

nit: collection

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionGet.json [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        "id": "/subscriptions/subid/providers/Microsoft.DocumentDb/locations/westus/restorableDatabaseAccounts/98a570f2-63db-4117-91f0-366327b7b353/restorableMongodbCollections/79609a98-3394-41f8-911f-cfab0c075c86",
+        "type": "Microsoft.DocumentDB/locations/restorableDatabaseAccounts/restorableMongodbCollections",
+        "name": "79609a98-3394-41f8-911f-cfab0c075c86",
+        "properties": {
+          "resource": {
+            "_rid": "zAyAPQAAAA==",
+            "eventTimestamp": "2020-10-13T04:56:42Z",
+            "ownerId": "Collection1",
+            "ownerResourceId": "V18LoLrv-qA=",
+            "operationType": "Create"
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionGet.json:21]: Not returning the MongoDBCollection in here ??

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            "type": "Microsoft.DocumentDB/locations/restorableDatabaseAccounts/restorableMongodbCollections",
+            "name": "79609a98-3394-41f8-911f-cfab0c075c86",
+            "properties": {
+              "resource": {
+                "_rid": "zAyAPQAAAA==",
+                "eventTimestamp": "2020-10-13T04:56:42Z",
+                "ownerId": "Collection1",
+                "ownerResourceId": "V18LoLrv-qA=",
+                "operationType": "Create"
+              }
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json:24]: same comment: Not returning the MongoDBCollection in here ??

**Review by MehaKaushik — CHANGES_REQUESTED** [review]
Left some comments

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionGet.json (hunk 13 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionGet.json:21]: Due to the different ways of getting the Collection content in Mongo 3.2 vs 3.6, we have decided to not include the content at this time.  The crucial infos (ie. operationtype, ownerId, and ownerResourceId) are still available.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json (hunk 14 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json:24]: We are not including the content for now.  Please see the other comment for details.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 8 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:8388]: sure. made it enum

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlResourceList.json (hunk 6 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlResourceList.json:16]: The restore api takes in a generic DatabaseRestoreResource object that's not API specific, and we are using the same object definition for both SQL and MongoDB here, so we don't have distinction between collection/container.

This will change once we update the restore api to takes in API specific parameters

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseList.json (hunk 9 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlDatabaseList.json:43]: For Mongo, we don't have the "Content" properties, so we need to expose it here.  Doing the same for Sql APIs for consistency

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json (hunk 11 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json:21]: For now, I think it is okay to keep it as is since we have the description and also the value ("Container1") should be indicative that's it's a collection/container name.
We can improve on the name in the next version, since this name is currently getting deployed.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json (hunk 5 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableMongodbCollectionList.json:7]: spoke offline.  will keep as-is since we need to distinguish between different objects

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 4 above)
Comment (vinhms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json]: moved to a separate restorable.json file

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/subid/providers/Microsoft.DocumentDb/locations/westus/restorableDatabaseAccounts/98a570f2-63db-4117-91f0-366327b7b353/restorableSqlContainers/79609a98-3394-41f8-911f-cfab0c075c86",
+            "type": "Microsoft.DocumentDB/locations/restorableDatabaseAccounts/restorableSqlContainers",
+            "name": "79609a98-3394-41f8-911f-cfab0c075c86",
+            "properties": {
+              "resource": {
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/examples/CosmosDBRestorableSqlContainerList.json:18]: I thought we not returning the original resource object anymore.

**Review by MehaKaushik — APPROVED** [review]
LGTM. Few things:
1. Please change the title of the PR. 
2. Are we having two different behaviors for Mongo Collection and Sql container LIST apis. In Sql we are returning the resource as well, but in Mongo we are not. I would recommend to keep both of the behaviors same.
3. For sanity check, please download any one of the SDKs (python, go, java) and have a look at the generated code. If anything is missed or not as expected we can change right away.
